File: topn_test.go
package dopdb

import (
	"fmt"
	"math/rand"
	"sort"
	"testing"
)

// The bounded retention must return EXACTLY what a full sort-then-slice would.
// It is an optimisation, so any difference is a correctness bug, not a trade-off.
func TestTopNMatchesFullSort(t *testing.T) {
	orders := [][]SortKey{
		nil,
		{{Field: "age", Asc: true}},
		{{Field: "age", Asc: false}},
		{{Field: "age", Asc: true}, {Field: "name", Asc: false}},
		{{Field: "_id", Asc: false}},
	}
	for _, keys := range orders {
		for _, n := range []int{1, 3, 10, 50, 500} {
			rows := make([]row, 0, 300)
			for i := 0; i < 300; i++ {
				rows = append(rows, row{
					id:  fmt.Sprintf("k%03d", i),
					doc: map[string]any{"age": int64(rand.Intn(20)), "name": fmt.Sprintf("n%03d", rand.Intn(50))},
				})
			}
			order := rowOrder{keys: keys}

			// reference: sort everything, take the first n
			want := make([]row, len(rows))
			copy(want, rows)
			sort.SliceStable(want, func(i, j int) bool { return order.less(want[i], want[j]) })
			if len(want) > n {
				want = want[:n]
			}

			h := newTopN(n, order)
			for _, r := range rows {
				h.push(r)
			}
			got := h.sorted()

			if len(got) != len(want) {
				t.Fatalf("keys=%v n=%d: got %d rows want %d", keys, n, len(got), len(want))
			}
			for i := range want {
				if got[i].id != want[i].id {
					t.Fatalf("keys=%v n=%d: row %d = %s want %s", keys, n, i, got[i].id, want[i].id)
				}
			}
		}
	}
}

File: ts/test/topn.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { TopN, rowLess, type Doc, type SortKey } from "../src/query.js";

// Bounded retention must return EXACTLY what a full sort-then-slice would. It is
// an optimisation, so any difference is a correctness bug, not a trade-off.
// (An inverted heap comparison here kept the WORST n instead of the best, and
// only the end-to-end tests caught it — hence this direct one.)
test("TopN matches a full sort followed by slice", () => {
  const orders: SortKey[][] = [
    [],
    [{ field: "age", asc: true }],
    [{ field: "age", asc: false }],
    [{ field: "age", asc: true }, { field: "name", asc: false }],
    [{ field: "_id", asc: false }],
  ];
  for (const keys of orders) {
    for (const n of [1, 3, 10, 50, 500]) {
      const rows: Doc[] = [];
      for (let i = 0; i < 300; i++) {
        rows.push({
          _id: `k${String(i).padStart(3, "0")}`,
          age: Math.floor(Math.random() * 20),
          name: `n${String(Math.floor(Math.random() * 50)).padStart(3, "0")}`,
        });
      }
      const want = [...rows]
        .sort((a, b) => (rowLess(a, b, keys) ? -1 : rowLess(b, a, keys) ? 1 : 0))
        .slice(0, n)
        .map((r) => r._id);

      const heap = new TopN(n, keys);
      for (const r of rows) heap.push(r);
      const got = heap.sorted().map((r) => r._id);

      assert.deepEqual(got, want, `keys=${JSON.stringify(keys)} n=${n}`);
    }
  }
});
